### ISSUE TITLE
LogSoftmax as a non-linearity

### ISSUE THREAD

**Issue by sidps** [Lasagne/Lasagne#332]
The Softmax non-linearity can lead to potential instabilities during training, LogSoftmax (i.e. log(Softmax(x)) ) seems to alleviate the issue. See #308 and [Theano/Theano#3162](https://github.com/Theano/Theano/issues/3162).

It might be useful to include LogSoftmax as a non-linearity and a corresponding Cross-entropy objective that assumes output probabilities are already in the log domain.

**Comment by f0k** [comment]
Thanks for the report! I think this should better be solved on the Theano side, so all users benefit from it and we don't need a lengthy explanation of what to do. If integration in Theano takes too long or turns out to be impossible, we should re-think this. In any case, please feel free to post your proposed nonlinearity function and objective function here so people can copy/paste it into their code if they can't wait for the issue being resolved upstream!

**Comment by sidps** [comment]
Sounds good. I am not using this at the moment, but the other person's code snippets are available in the Theano issue linked previously.

**Comment by f0k** [comment]
Okay, just for easier copy/pasting, the functions would be (untested):
```python
def log_softmax(x):
    xdev = x - x.max(1, keepdims=True)
    return xdev - T.log(T.sum(T.exp(xdev), axis=1, keepdims=True))

def categorical_crossentropy_logdomain(log_predictions, targets):
    return -T.sum(targets * log_predictions, axis=1)
```
Note that the latter doesn't support `targets` being an int vector now, that would be `log_predictions[targets]` instead.

**Comment by MInner** [comment]
Hey, guys, I guess you should add a note about it somewhere [here](https://lasagne.readthedocs.org/en/latest/modules/objectives.html#lasagne.objectives.categorical_crossentropy) :) And you also seem to have a really huge default learning rate (=1) for lasagne.updates.rmsprop , I've noticed people setting it to be 0.1 - 0.001 by default

**Comment by f0k** [comment]
> Hey, guys, I guess you should add a note about it somewhere here :)

Well, we're hoping for this to be fixed for good in Theano, without requiring any changes in Lasagne. It seems it's almost in there: https://github.com/Theano/Theano/pull/3363

> And you also seem to have a really huge default learning rate (=1) for lasagne.updates.rmsprop

We should probably document that. Just changing it to something else would break backwards compatibility (although maybe nobody's using it with 1.0 anyway). Thanks!

**Comment by nouiz** [comment]
Changing the 1.0 to have no default could be a good alternative. This way,
people will know what is there. If this make someone code crash, he will
probably be happy to know about this in his code.

**Comment by rparrapy** [comment]
I'm currently facing the nan gradient issue with a dataset with a lot of one-hot coded labels, so I'm trying to use the copy/pasted functions from @f0k.

`targets` is an int vector now and, as expected, the function does not work. Could you please clarify what the expected value for `targets` in the loss function is? I did not understand the `log_predictions[targets]` comment.

**Comment by f0k** [comment]
> I did not understand the `log_predictions[targets]` comment.

If `targets` is an int vector instead of a one-hot matrix, then `-T.sum(targets * log_predictions, axis=1)` should be replaced by `-log_predictions[T.arange(len(targets)), targets]`. This will pick the column indicated by `targets` for each row.

**Comment by rslprpr** [comment]
I have the same  problem and I've tried to use 
`def log_softmax(x):
    xdev = x - x.max(1, keepdims=True)
    return xdev - T.log(T.sum(T.exp(xdev), axis=1, keepdims=True))

def categorical_crossentropy_logdomain(log_predictions, targets):
    return -T.sum(targets * log_predictions, axis=1)`
But it says :  `object has no attribute 'max'`

Any idea for replacement?

**Comment by f0k** [comment]
> Any idea for replacement?

How did you use `log_softmax`? This is meant to be used as a layer's nonlinearity, as in `DenseLayer(layer, num_units=10, nonlinearity=log_softmax)`. Does this help?

If you're just generally fighting NaNs with categorical cross-entropy, it may also help to clip the predictions before passing them to the loss function:
```python
predictions = lasagne.layers.get_output(net, deterministic=False)
predictions = T.clip(predictions, 1e-7, 1 - 1e-7)
loss = lasagne.objectives.categorical_crossentropy(predictions, targets).mean()
```

**Comment by rslprpr** [comment]
Thanks, but the current version of Theano doesn't have`log_softmax` as nonlinearity.

**Comment by nouiz** [comment]
Use Theano dev version.

**Comment by f0k** [comment]
> Thanks, but the current version of Theano doesn't have`log_softmax` as nonlinearity.

The `def log_softmax(...)` snippet you quoted *is* a nonlinearity. No need to integrate it in Theano, just copy it into your code and use it via the `nonlinearity=` argument of a layer.

> Use Theano dev version.

Ah, hadn't seen that. The Theano dev version has `T.nnet.logsoftmax()`, this also works and might be preferable.

**Comment by nouiz** [comment]
I don't know what the lasagne version does, but the Theano version assert
that you will use a numerical stable version and will use cudnn version.

**Comment by f0k** [comment]
Lasagne exposes the standard `T.nnet.softmax` as `lasagne.nonlinearities.softmax`. This thread here was started because it lead to numeric instabilities in some cases, and manually using a log-softmax was a feasible workaround. I think this has been solved on the Theano side now.
@rslprpr, can you try updating to the Theano bleeding-edge version (http://lasagne.readthedocs.io/en/latest/user/installation.html#bleeding-edge-version) and see if this fixes your problems?

**Comment by rslprpr** [comment]
I updated both Lasagne and Theano to bleeding-edge version, added log softmax along with clipping network's output. But `nan` problem is still there. No idea how to fix it!

**Comment by f0k** [comment]
> But `nan` problem is still there. No idea how to fix it!

There can be many different causes. Try reducing the learning rate, normalizing your inputs, simplifying the network architecture, using batch normalization (`lasagne.layers.batch_norm` or `lasagne.layers.dnn.batch_norm_dnn`). There's also a section in the Theano documentation: http://deeplearning.net/software/theano_versions/dev/tutorial/nan_tutorial.html

**Comment by rslprpr** [comment]
You were right. I've tried all the solutions you said before and non of them worked till I finally figured out that few images between my 100million image dataset which have cut and it means the those images were passing NAN value to the network! .. I eliminate those and now everything works fine. Thanks for help!

**Comment by rafaelvalle** [comment]
Can anyone explain why Lasagne, Theano, Numpy, etc allow mathematical operations between numbers and nans?

**Comment by f0k** [comment]
> Can anyone explain why Lasagne, Theano, Numpy, etc allow mathematical operations between numbers and nans?

Because the processor instruction sets allow it, and they're not necessarily evil -- e.g., you can use NaNs to mark invalid input values and they'll propagate through any element-wise operations.

I'm not sure what you'd want it to do instead, but you can configure numpy to print a warning or raise an exception when an operation results in a NaN: https://docs.scipy.org/doc/numpy/reference/generated/numpy.seterr.html. This won't help for Theano, though. Theano expressions can be run in numpy, compiled for CPU or for GPU (CUDA or OpenCL), and at least CUDA does not support trap handlers as on CPU (http://docs.nvidia.com/cuda/floating-point/#differences-from-x86). So in Theano, the only way is to check for NaNs in between operations, e.g., using the NaNGuardMode (http://deeplearning.net/software/theano/library/compile/nanguardmode.html), which is expensive and thus not done by default. If you just want to stop training on NaNs, you can simply adapt your training loop to exit if the error for a mini-batch is NaN, using `np.isnan()` or `not np.isfinite()`. This won't tell you where the NaN originated, but at least avoid wasting computation time.

**Comment by jingriver** [comment]
2016年10月21日星期五，Rafael Valle <notifications@github.com> 写道：

**Comment by rafaelvalle** [comment]
Can you, or anyone for that matter, describe why the rslprpr's network
output NaNs, even though he tried using logsoftmax, when computing the
categorical cross-entropy using a mini-batch(?) that included certain
images?

**Comment by f0k** [comment]
> Can you, or anyone for that matter, describe why the rslprpr's network output NaNs

I think only @rslprpr can tell. I didn't understand what she meant by "few images between my 100million image dataset which have cut and it means the those images were passing NAN value" -- @rslprpr, can you elaborate what was wrong with those images to enlighten @rafaelvalle?

**Comment by rslprpr** [comment]
I've tried all the possible solution that @f0k mentioned in previous comments but the output of network still was NaNs. After couple of days I figured out in my dataset(which is a really big dataset containing 100 millions images) I have number of images with NaNs values, it means when I cut the 28*28 patches from movie frame, some patches contain less than 28*28 areas so it considered the remaining area in the patch as NaNs. No matter what exactly you do in your network when the input is Nan, the output will be Nan! I eliminated those patches and the network works fine!
